### PR TITLE
Add support for atomic instructions in bpf2c

### PR DESCRIPTION
**Pull request by b1tg**
## Description

Support for atomic instructions in bpf2c 

Fixes: #1819 

## Testing

Use bpf_conformance_runner.exe to test on my local environment, result:
```
> bpf_conformance_runner.exe --test_file_directory D:\ebpf-for-windows\external\ubpf\external\bpf_conformance\tests\ --plugin_path .\bpf2c_plugin.exe --plugin_options "--include D:\ebpf-for-windows\include"
[...]
FAIL: "D:\\ebpf-for-windows\\external\\ubpf\\external\\bpf_conformance\\tests\\call_local.data" Plugin returned error code 1 and output bpf_test doesn't support helper id=9
[...]
Passed 136 out of 137 tests.
```
I found call_local.data failed under brach main too.

## Documentation

`docs\isa-support.rst` was updated

### CODE REVIEW

File: include/bpf2c.h
Comment (dv-msft) [inline include/bpf2c.h:6]: Couple of queries:

1. Why are we using this header (from clang) when we have windows implementation for the ```Interlocked...``` family in ```winnt.h```?
2. Why is the header being included here? Given its specificity to the ```bpf_code_generator.cpp``` source file, maybe it could be included directly in that source itself?  That would localize the header dependency to the source(s) where it's actually used.

File: tools/bpf2c/bpf_code_generator.cpp
Comment (dv-msft) [inline tools/bpf2c/bpf_code_generator.cpp:742]: Is there a reason why we could not use the windows version of [InterlockedAdd](https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedadd) (and for other interlocked intrinsics used in this PR) instead? Windows has equivalents for all the clang ```_InterLocked...``` intrinsics used here.

That might also do away with the necessity of having to include clang's ```intrin.h```

Comment (b1tg) [reply tools/bpf2c/bpf_code_generator.cpp:742]: `instrin.h` is also in MSVC (C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.33.31629\include\intrin.h). As I know, we use cl.exe to build generated code in this project rather than clang, so using `_InterLocked*` here has no dependency on clang.

Using `_Interlocked*` or `Interlocked*` is both ok, one include `intrin.h`, another include `winnt.h`.

Comment (dv-msft) [reply tools/bpf2c/bpf_code_generator.cpp:742]: AFAIK we prefer the ```Interlocked...``` versions from ```winnt.h``` at least insofaras windows specific components are concerned.  That being said, I'm relatively new to this project as well, so I'll defer to the veterans on this one :-)

@Alan-Jowett, @dthaler: Thoughts?

File: include/bpf2c.h
Comment (b1tg) [reply include/bpf2c.h:6]: If we want atomic function header only accurs when the file call those functions, we may need a flag to indicate
if atomic function was used, change like this:

```cpp
// bpf_code_generator.cpp
void
bpf_code_generator::encode_instructions(const bpf_code_generator::unsafe_string& section_name)
{
// [...]
            if ((inst.opcode & EBPF_MODE_MASK) == EBPF_MODE_ATOMIC) { // MODE_ATOMIC
                need_atomic = true; // set flag
                auto line = std::string("");
// [...]
void
bpf_code_generator::emit_c_code(std::ostream& output_stream)
{
    // Emit C file
    output_stream << "#include \"bpf2c.h\"" << std::endl << std::endl;
    if (need_atomic) {
        output_stream << "#include <intrin.h>" << std::endl << std::endl;
    }
```

And I just found those atomic functions are also defined in `wdm.h`, so for `*_sys.c`, we do not need to include another header. If we can tolerate include intrin.h in bpf2c.h, we can just add intrin.h in bpf2c.h like this:

```cpp
// bpf2c.h
#if defined(NO_CRT)
//[...]
#else
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <intrin.h>
#endif
```

Which way is better?

Comment (dv-msft) [reply include/bpf2c.h:6]: That's the thing :-) What's confusing me why we even _need_ ```intrin.h``` when we have native windows equivalents (via native windows headers).  I'm quite possibly missing something here. In any case it looks like this involves some 'policy' issues so I'd like to hear what @Alan-Jowett / @dthaler have to say about this.

We can discuss this issue during our Monday morning meeting.

**Comment by Alan-Jowett** [comment]
Would you kindly add tests samples that use these instructions and updating the bpf2c tests?
bpf2c tests:
https://github.com/microsoft/ebpf-for-windows/blob/eb83ea88c297dbfe0c5bec6feb77c07fa418a9ce/tests/bpf2c_tests/elf_bpf.cpp#L171

That will make verifying that the emitted C code is correct much more easily as it includes both the ELF file and the generated c files.

File: tools/bpf2c/bpf_code_generator.cpp
Comment (b1tg) [reply tools/bpf2c/bpf_code_generator.cpp:742]: Any update? @dv-msft

Comment (dv-msft) [reply tools/bpf2c/bpf_code_generator.cpp:742]: Could you please add the tests recommended by @Alan-Jowett in his comment? That emitted C code output might give us better picture.

File: include/bpf2c.h
Comment (Alan-Jowett) [reply include/bpf2c.h:6]: Can we get a sample that uses these instructions added first? I would like to see what the generated code looks like before I comment.

**Comment by b1tg** [comment]
bpf2c test sample added. @Alan-Jowett @dv-msft

File: tests/bpf2c_tests/expected/atomic_instruction_raw.c
Comment (Alan-Jowett) [inline tests/bpf2c_tests/expected/atomic_instruction_raw.c:110]: Can you update it to show the instruction being emulated in the debug info here?

File: tests/sample/atomic_instruction.c
Comment (Alan-Jowett) [inline tests/sample/atomic_instruction.c:25]: Is it possible to have this generate both the fetch and non-fetch version of the instructions? I want to make sure what we generate is correct for both cases.

Comment (Alan-Jowett) [reply tests/sample/atomic_instruction.c:25]: See:
https://github.com/dthaler/ebpf-docs/blob/update/isa/kernel.org/instruction-set.rst#152atomic-operations

image

Comment (Alan-Jowett) [reply tests/sample/atomic_instruction.c:25]: Thanks to @dylandreimerink for providing the answer:
>Dylan Reimerink
  [1 hour ago](https://cilium.slack.com/archives/C4XCTGYEM/p1674753252369159?thread_ts=1674751865.973939&cid=C4XCTGYEM)
By using the return value or not.
The following example emits with a FETCH:
> #include <linux/types.h>

>
> #include "bpf_helpers.h"

>
>__u64 add64_value = 1;

>
>SEC("raw_tp/sys_enter")
>int add(const void *ctx)
>{
>    __u64 b = __sync_fetch_and_add(&add64_value, 2);

>
>    return b;
>}

>
>atomic: file format elf64-bpf

>
>Disassembly of section raw_tp/sys_enter:

>
>0000000000000000 <add>:
>       0:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0 ll
>       2:       61 16 00 00 00 00 00 00 r6 = *(u32 *)(r1 + 0)
>       3:       85 00 00 00 0e 00 00 00 call 14
>       4:       bf 01 00 00 00 00 00 00 r1 = r0
>       5:       b7 00 00 00 00 00 00 00 r0 = 0
>       6:       77 01 00 00 20 00 00 00 r1 >>= 32
>       7:       5d 61 04 00 00 00 00 00 if r1 != r6 goto +4 <LBB0_2>
>       8:       b7 00 00 00 02 00 00 00 r0 = 2
>       9:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0 ll
>      11:       db 01 00 00 01 00 00 00 r0 = atomic_fetch_add((u64 *)(r1 + 0), r0)

>
>0000000000000060 <LBB0_2>:
>      12:       95 00 00 00 00 00 00 00 exit
>This one emits without a FETCH:
>#include <linux/types.h>

>
>#include "bpf_helpers.h"

>
>__u64 add64_value = 1;

>
>SEC("raw_tp/sys_enter")
>int add(const void *ctx)
>{
>    __sync_fetch_and_add(&add64_value, 2);

>
>    return 2;
>}

>
>atomic: file format elf64-bpf

>
>Disassembly of section raw_tp/sys_enter:

>
>0000000000000000 <add>:
>       0:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0 ll
>       2:       b7 02 00 00 02 00 00 00 r2 = 2
>       3:       db 21 00 00 00 00 00 00 lock *(u64 *)(r1 + 0) += r2
>       4:       b7 00 00 00 02 00 00 00 r0 = 2
>       5:       95 00 00 00 00 00 00 00 exit

File: tests/bpf2c_tests/expected/atomic_instruction_raw.c
Comment (b1tg) [reply tests/bpf2c_tests/expected/atomic_instruction_raw.c]: fixed.

File: tests/sample/atomic_instruction.c
Comment (b1tg) [reply tests/sample/atomic_instruction.c]: 1. updated at https://github.com/microsoft/ebpf-for-windows/pull/1936/commits/731be45e5d2abaa8888c56f03b00a09e26e9790b
2. clang-11 doesn't recognize atomic instructions other than xadd, i use clang-15.0.6 to build atomic_instruction.c, and i notice we use clang-11.0.1 across this project, how do we handle this issue.
3. external\ebpf-verifier doesn't support atomic instruction yet, i temporarily add --no-verify in 
   user_mode_bpf2c.vcxproj and kernel_mode_bpf2c.vcxproj to make the build not fail.
4. seems even with clang15, when generating 32bit atomic instruction, only `__sync_fetch_and_add` works, other atomic functions like `__sync_fetch_and_sub` fail.

```
PS > type .\atomic.c
typedef unsigned long u64;
typedef unsigned int u32;
__attribute__((section("prog"), used))
int test( void *ctx)
{
    u32 aa = 7;
    __sync_fetch_and_sub(&aa, 2);
    return 0;
}
PS > clang -target bpf -c .\atomic.c
fatal error: error in backend: Cannot select: 0x1e805cfd678: i64,ch = AtomicLoadSub<(load store seq_cst (s32) on
      %ir.3)> 0x1e805cfd610, FrameIndex:i64<1>, Constant:i64<2>
  0x1e805cfd268: i64 = FrameIndex<1>
  0x1e805cfd408: i64 = Constant<2>
In function: test
[...]
PS > type .\atomic.c
typedef unsigned long u64;
typedef unsigned int u32;
__attribute__((section("prog"), used))
int test( void *ctx)
{
    u32 aa = 7;
    __sync_fetch_and_or(&aa, 2);
    return 0;
}
PS > clang -target bpf -c .\atomic.c
.\atomic.c:4:5: error: Unsupported atomic operations, please use 64 bit version
int test( void *ctx)
    ^
fatal error: error in backend: Cannot select: 0x1c474dc5f28: i64,ch = AtomicLoadOr<(load store seq_cst (s32) on       
      %ir.3)> 0x1c474dc5ec0, FrameIndex:i64<1>, Constant:i64<2>
  0x1c474dc5b18: i64 = FrameIndex<1>
  0x1c474dc5cb8: i64 = Constant<2>
In function: test

```

Comment (dthaler) [reply tests/sample/atomic_instruction.c]: Thanks for working on this and the great info!

> 2. clang-11 doesn't recognize atomic instructions other than xadd, i use clang-15.0.6 to build atomic_instruction.c, and i notice we use clang-11.0.1 across this project, how do we handle this issue.

For now I would just have tests that just use the byte code directly rather than compiling from C.  If compiling from C is absolutely needed for some reason, then maybe using asm() in the C to embed the unsupported ebpf instruction.  Using a later version of clang is tracked by issue #1111.

> 3. external\ebpf-verifier doesn't support atomic instruction yet, i temporarily add --no-verify in
>    user_mode_bpf2c.vcxproj and kernel_mode_bpf2c.vcxproj to make the build not fail.

Correct the verifier doesn't have support yet.  I would suggest holding off on bpf2c testing of atomic instructions until the verifier does support it.  Fine to manually test like you did but have atomics be skipped in cicd (I don't think we should let bpf2c support no-verify).

Comment (b1tg) [reply tests/sample/atomic_instruction.c]: Inline assembly is not recognized by clang11 either, shoud we leave only `__sync_fetch_and_add` in atomic_instruction.c
or just don't include test sample in this pr.

```
PS > type .\atomic.c
typedef unsigned long u64;
typedef unsigned int u32;
__attribute__((section("prog"), used))
int test( void *ctx)
{
    u64 a = 2, b = 3, *c = &a;
    asm volatile("lock *(u64 *)(%0+0) += %1" : "=r"(c) : "r"(b), "0"(c));
    asm volatile("lock *(u64 *)(%0+0) &= %1" : "=r"(c) : "r"(b), "0"(c));
    return a;
}
PS > clang -target bpf -c .\atomic.c # clang15
PS > & 'c:\Program Files\LLVM-11\bin\clang.exe'  -target bpf -c .\atomic.c #clang11
.\atomic.c:8:18: error: invalid operand for instruction
    asm volatile("lock *(u64 *)(%0+0) &= %1" : "=r"(c) : "r"(b), "0"(c));
                 ^
<inline asm>:1:22: note: instantiated into assembly here
        lock *(u64 *)(r2+0) &= r1
                            ^
1 error generated.
```

Comment (dthaler) [reply tests/sample/atomic_instruction.c]: > shoud we leave only __sync_fetch_and_add in atomic_instruction.c or just don't include test sample in this pr?

Leave only the atomic add for things compiled from C.
Other atomic tests can act on bytecode generated directly from tests rather than by clang.

Comment (b1tg) [reply tests/sample/atomic_instruction.c]: 8693960ac5ff: Split the atomic test into two test cases

Current ci was failed because i changed count increment in `tests/sample/droppacket.c` to atomic add and unit_test failed because [ubpf not support atomic instructions](https://github.com/iovisor/ubpf/issues/64) yet, should i remove this change or find a workaround?
https://github.com/microsoft/ebpf-for-windows/blob/8693960ac5ff0ef574959d5efd8bd05fa7fc3044/tests/sample/droppacket.c#L71-L72
Are there any other things I missed or did not do right?

Comment (b1tg) [reply tests/sample/atomic_instruction.c]: And for reference, here is the atomic_instruction_others.o disassembly output (llvm-objdump.exe LLVM version 15.0.6):
```
PS > llvm-objdump.exe -d ebpf-for-windows\tests\sample\atomic_instruction_others.o

ebpf-for-windows\tests\sample\atomic_instruction_others.o:     file format elf64-bpf

Disassembly of section xdp_prog:

0000000000000000 <test>:
       0:       7b 1a f8 ff 00 00 00 00 *(u64 *)(r10 - 8) = r1
       1:       b7 02 00 00 7b 00 00 00 r2 = 123
       2:       7b 2a f0 ff 00 00 00 00 *(u64 *)(r10 - 16) = r2
       3:       b7 00 00 00 03 00 00 00 r0 = 3
       4:       7b 0a e8 ff 00 00 00 00 *(u64 *)(r10 - 24) = r0
       5:       b7 01 00 00 00 00 00 00 r1 = 0
       6:       7b 1a c8 ff 00 00 00 00 *(u64 *)(r10 - 56) = r1
       7:       7b 1a e0 ff 00 00 00 00 *(u64 *)(r10 - 32) = r1
       8:       63 2a dc ff 00 00 00 00 *(u32 *)(r10 - 36) = r2
       9:       63 0a d8 ff 00 00 00 00 *(u32 *)(r10 - 40) = r0
      10:       63 1a d4 ff 00 00 00 00 *(u32 *)(r10 - 44) = r1
      11:       b7 01 00 00 01 00 00 00 r1 = 1
      12:       bf 12 00 00 00 00 00 00 r2 = r1
      13:       db 2a f0 ff 01 00 00 00 r2 = atomic_fetch_add((u64 *)(r10 - 16), r2)
      14:       b7 02 00 00 02 00 00 00 r2 = 2
      15:       db 2a f0 ff 41 00 00 00 r2 = atomic_fetch_or((u64 *)(r10 - 16), r2)
      16:       bf 02 00 00 00 00 00 00 r2 = r0
      17:       db 2a f0 ff 51 00 00 00 r2 = atomic_fetch_and((u64 *)(r10 - 16), r2)
      18:       b7 02 00 00 04 00 00 00 r2 = 4
      19:       db 2a f0 ff a0 00 00 00 lock *(u64 *)(r10 - 16) ^= r2
      20:       c3 1a dc ff 40 00 00 00 <unknown>
      21:       79 a1 e8 ff 00 00 00 00 r1 = *(u64 *)(r10 - 24)
      22:       db 1a f0 ff e1 00 00 00 r1 = xchg_64(r10 - 16, r1)
      23:       7b 1a e0 ff 00 00 00 00 *(u64 *)(r10 - 32) = r1
      24:       b7 01 00 00 05 00 00 00 r1 = 5
      25:       db 1a f0 ff f1 00 00 00 r0 = cmpxchg_64(r10 - 16, r0, r1)
      26:       79 a0 c8 ff 00 00 00 00 r0 = *(u64 *)(r10 - 56)
      27:       95 00 00 00 00 00 00 00 exit
```

File: include/bpf2c.h
Comment (dthaler) [inline include/bpf2c.h:24]: Move include up into alphabetical order per [DevelopmentGuide](https://github.com/microsoft/ebpf-for-windows/blob/main/docs/DevelopmentGuide.md#header-files).

Comment (b1tg) [reply include/bpf2c.h]: Fixed.

**Review by dthaler — APPROVED** [review]

**Review by dthaler — COMMENTED** [review]
Just realized: please update the bpf2c column of docs\isa-support.rst for the instructions now supported.  And update the PR description's Documentation section to mention that docs were updated.

File: docs/isa-support.rst
Comment (dthaler) [inline docs/isa-support.rst:126]: Only the bpf2c column should be updated, since PREVAIL and uBPF do not support these instructions.

**Review by dthaler — APPROVED** [review]
Looks ok to me now, thanks!

**Review by Alan-Jowett — DISMISSED** [review]

**Comment by dthaler** [comment]
@b1tg red can you resolve merge conflicts?  thanks!

**Comment by dthaler** [comment]
@b1tg tests failed because version number bumped so you need to regenerate the expected bpf2c output. i.e.:
```
Expected:     version->minor = 5;
Actual:     version->minor = 6;
```

**Comment by dthaler** [comment]
@b1tg please resolve conflicts, would like to get this merged :)

**Pull request merged**
